src/coreclr/src/vm/tailcallhelp.cpp

```diff
@@ -447,42 +476,31 @@ MethodDesc* TailCallHelp::CreateCallTargetStub(const TailCallInfo& info)
     auto emitOffs = [&](UINT offs)
     {
         pCode->EmitLDARG(ARG_ARG_BUFFER);
-        if (offs != 0)
-        {
-            pCode->EmitLDC(offs);
-            pCode->EmitADD();
-        }
+        pCode->EmitLDC(offs);
+        pCode->EmitADD();
     };
 
-    StackSArray<DWORD> argLocals;
+    // *pTailCallAwareRetAddr = NextCallReturnAddress();
+    pCode->EmitLDARG(ARG_PTR_TAILCALL_AWARE_RET_ADDR);
+    pCode->EmitCALL(METHOD__STUBHELPERS__NEXT_CALL_RETURN_ADDRESS, 0, 1);
+    pCode->EmitSTIND_I();
+
     for (COUNT_T i = 0; i < info.ArgBufLayout.Values.GetCount(); i++)
     {
         const ArgBufferValue& arg = info.ArgBufLayout.Values[i];
-        DWORD argLcl = pCode->NewLocal(LocalDesc(arg.TyHnd));
-        argLocals.Append(argLcl);
 
         // arg = args->Arg_i
         emitOffs(arg.Offset);
         EmitLoadTyHnd(pCode, arg.TyHnd);
-        pCode->EmitSTLOC(argLcl);
-    }
-
-    DWORD targetAddrLcl;
-    if (info.ArgBufLayout.HasTargetAddress)
-    {
-        targetAddrLcl = pCode->NewLocal(ELEMENT_TYPE_I);
-
-        emitOffs(info.ArgBufLayout.TargetAddressOffset);
-        pCode->EmitLDIND_I();
-        pCode->EmitSTLOC(targetAddrLcl);
     }
 
-    // RuntimeHelpers.FreeTailCallArgBuffer();
-    pCode->EmitCALL(METHOD__RUNTIME_HELPERS__FREE_TAILCALL_ARG_BUFFER, 0, 0);
-
-    // *pTailCallAwareRetAddr = NextCallReturnAddress();
-    pCode->EmitLDARG(ARG_PTR_TAILCALL_AWARE_RET_ADDR);
-    pCode->EmitCALL(METHOD__STUBHELPERS__NEXT_CALL_RETURN_ADDRESS, 0, 1);
+    // All arguments are loaded on the stack, it is safe to disable the GC reporting of ArgBuffer now.
+    // This is optimization to avoid extending argument lifetime unnecessarily.
+    // We still need to report the inst argument of shared generic code to prevent it from being unloaded. The inst
```

Comment:
FYI: In theory, I have noticed that nothing guarantees that the tailcall target function pointer is kept alive and maybe unloaded. It would be a very corner-case situation that is unlikely to be ever hit in a real-world. I have opened #41314 on it.